### PR TITLE
[th2-1929] Fix problem with retry grpc call when a server-side thrown…

### PR DESCRIPTION
**Pull request by Nikita-Smirnov-Exactpro**
… exception

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,6 +61,10 @@ protobuf {
 
 ## Release notes
 
+### 3.1.11
+
+* Fix problem with retry grpc call when a server-side thrown exception
```
Comment (jean-michel-perli-sonsin) [inline README.md:66]:
```suggestion
* Fix a problem with retry grpc call when a server-side thrown an exception
```

File: README.md [hunk 2]
```diff
@@ -61,6 +61,10 @@ protobuf {
...
+
 ### 3.1.10
 
 * Fix bug with using incorrect message if message with same name exist in different proto package;
```
Comment (jean-michel-perli-sonsin) [inline README.md:70]:
```suggestion
* Fix a bug which uses an incorrect message, if a message with the same name exists in a different proto package;
```

**Review by jean-michel-perli-sonsin — CHANGES_REQUESTED** [review]
Hello, please see if you agree with the changes proposed below:
- README.md
line 4: You can plug it **directly** to protoc or run it from the command line **itself**.

line 8: **It is possible to** set the following options:

line 47: You can run this project using **the** command line. In order to this**, you need to** refer to the project's artifact which is used in Gradle protobuf's plugin.

line 49: **The first** argument should be **the directory or the file** with ``FileDescriptorSet``. **The second** argument is **the** output directory. After **those arguments,** you can set **the** options (Example: `--pythonPath=./path/to/python`)

line 51: ``FileDescriptorSet`` is a special protobuf's object **which describes the** proto files. You can generate it with Gradle's protobuf plugin

- src/main/java/com/exactpro/th2/service/AbstractGrpcService.java
line 2: Copyright 2020-**2021** Exactpro (Exactpro Systems Limited)

In case of any question, please let me know.
Thanks!

**Review by Ilya-Loginov-Exactpro — DISMISSED** [review]
I think we should check this case in `RetryStreamObserver` in method `onError`

File: src/main/java/com/exactpro/th2/service/AbstractGrpcService.java [hunk 3]
```diff
@@ -111,14 +115,18 @@ public void onError(Throwable t) {
             int attempt = currentAttempt.getAndIncrement();
             if (attempt < retryPolicy.getMaxAttempts() && t instanceof StatusRuntimeException) {
 
-                logger.warn("Can not send GRPC async request. Retrying. Current attempt = {}", currentAttempt.get() + 1, t);
+                if (((StatusRuntimeException)t).getStatus() == Status.UNKNOWN) { // Server side thrown an exception
```
Comment (Ilya-Loginov-Exactpro) [inline src/main/java/com/exactpro/th2/service/AbstractGrpcService.java:118]: Are we sure `t` is alwayse `StatusRuntimeException`? I think we should check it.

**Review by Ilya-Loginov-Exactpro — CHANGES_REQUESTED** [review]

**Review by jean-michel-perli-sonsin — APPROVED** [review]

File: src/main/java/com/exactpro/th2/service/AbstractGrpcService.java (hunk 3 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/java/com/exactpro/th2/service/AbstractGrpcService.java:118]: We checked it in the if block
if (attempt < retryPolicy.getMaxAttempts() && **t instanceof StatusRuntimeException**) {

**Review by Ilya-Loginov-Exactpro — APPROVED** [review]

**Pull request merged**